Sources/NIO/AddressedEnvelope.swift

```diff
@@ -20,15 +20,35 @@
 public struct AddressedEnvelope<DataType> {
     public var remoteAddress: SocketAddress
     public var data: DataType
+    /// Any metadata associated with this `AddressedEnvelope`
+    public var metadata : Metadata? = nil
 
     public init(remoteAddress: SocketAddress, data: DataType) {
         self.remoteAddress = remoteAddress
         self.data = data
     }
+    
+    /// Any metadata associated with an `AddressedEnvelope`
+    public struct Metadata {
+        /// Details of any congestion state.
+        public var ecnState : NIOECNState
+    }
 }
 
 extension AddressedEnvelope: CustomStringConvertible {
     public var description: String {
         return "AddressedEnvelope { remoteAddress: \(self.remoteAddress), data: \(self.data) }"
     }
 }
+
+/// Possible ECN States
+public enum NIOECNState: Hashable {
+    /// Non-ECN Capable Transport.
+    case nonECT
+    /// ECN Capable Transport (flag 0).
+    case ect0
+    /// ECN Capable Transport (flag 1).
+    case ect1
+    /// Congestion Experienced.
+    case ce
```

Comment:
nit: `congestionEncountered`/`congestionExperienced`

Reply:
I was heavily influenced by
https://developer.apple.com/documentation/network/nwprotocolip/ecn which has experienced.
Wikipedia https://en.wikipedia.org/wiki/Explicit_Congestion_Notification goes for encountered.
RFC https://tools.ietf.org/html/rfc3168 goes for experienced.

I think experienced wins 2:1 :-)

Reply:
I think George is proposing changing the name of the enum case. ;)

Reply:
Correct. We should spell it out either way to gain some clarity here.

Reply:
Ahh - I should probably spell them all out then.

Reply:
I have expanded the naming in this area.